### PR TITLE
Add API to edit and delete an add-on user review

### LINKED ISSUE #3401: Add API to edit and delete an add-on review
Splitted from #3342, supports #3024

We need an API to edit and delete add-on reviews. Review authors or admins have access to this.

### PR DESCRIPTION
**Pull request by diox**
Fix #3401

### CODE REVIEW

File: src/olympia/reviews/serializers.py [hunk 1]
```diff
@@ -53,6 +53,10 @@ def to_representation(self, obj):
         return data
 
     def validate_version(self, version):
+        if self.partial:
+            raise serializers.ValidationError(
```
Comment (eviljeff) [inline src/olympia/reviews/serializers.py]: test for this validationError?

File: src/olympia/reviews/serializers.py [hunk 2]
```diff
@@ -66,12 +70,13 @@ def validate(self, data):
...
-        if Review.objects.filter(
-                addon=data['addon'], user=data['user'],
-                version=data['version']).exists():
-            raise serializers.ValidationError(
-                'The same user can not leave a review on the same version more'
-                ' than once.')
+        if not self.partial:
+            if Review.objects.filter(
+                    addon=data['addon'], user=data['user'],
+                    version=data['version']).exists():
```
Comment (EnTeQuAk) [inline src/olympia/reviews/serializers.py]: I really hate if-statements like that, can we do something like ``review_exists = Review.objects...`` and ``if review_exists: ...``?

File: src/olympia/reviews/serializers.py (hunk 1 above)
Comment (diox) [inline src/olympia/reviews/serializers.py]: it's tested in the view tests

File: src/olympia/reviews/views.py [hunk 3]
```diff
@@ -343,7 +340,10 @@ class ReviewViewSet(AddonChildMixin, CreateModelMixin, ListModelMixin,
 
             # To edit a review you need to be the author or be an admin.
             'patch': AnyOf(AllowOwner, GroupPermission('Addons', 'Edit')),
-            'put': AnyOf(AllowOwner, GroupPermission('Addons', 'Edit')),
+
+            # Implementing PUT would be a little incoherent as we don't want to
+            # allow users to change `version` but require it at creation time.
+            # So only PATCH is allowed for editing.
```
Comment (EnTeQuAk) [inline src/olympia/reviews/views.py]: 👍

**Comment by EnTeQuAk** [comment]
r+wc

File: docs/topics/api/reviews.rst [hunk 4]
```diff
@@ -81,6 +81,43 @@ This endpoint allows you to post a new review for a given add-on and version.
...
+     Requires authentication and admin permissions or the user account that
+     posted the review.
+
+     Only body, title and rating are allowed for modification.
+
+.. http:patch:: /api/v3/addons/addon/(int:id|string:slug|string:guid)/reviews/(int:id)/
+
+    :<json string|null body: The text of the review.
+    :<json string|null: The title of the review.
+    :<json int rating: The rating the user wants to give as part of the review (required).
```
Comment (eviljeff) [inline docs/topics/api/reviews.rst]: Is rating still required for an edit?  Seems like it would have been given during the initial submission.

Comment (diox) [inline docs/topics/api/reviews.rst]: No, that's a bad copy/paste in the docs. Fixed.

File: src/olympia/reviews/serializers.py (hunk 2 above)
Comment (diox) [inline src/olympia/reviews/serializers.py]: Done

**Pull request merged**